upstairs/src/volume.rs

```diff
@@ -3549,76 +3630,116 @@ mod test {
             panic!("wrong variant returned!");
         };
 
-        info!(log, "replace {old} with {new}");
+        info!(log, "replaced {old} with {new}");
         assert_eq!(original_target, old);
         assert_eq!(new_target, new);
         Ok(())
     }
 
     #[test]
-    fn volume_replace_rop() {
-        // A replacement VCR is provided with one target being
-        // different, both new and old have a read_only_parent
+    fn test_volume_replace_no_gen() {
+        // We need at least two sub_volumes for this test.
+        for sv in 2..4 {
+            for sv_changed in 0..sv {
+                test_volume_replace_no_gen_inner(sv, sv_changed, false);
+                test_volume_replace_no_gen_inner(sv, sv_changed, true);
+            }
+        }
+    }
+
+    fn test_volume_replace_no_gen_inner(
+        sv_count: usize,
+        sv_changed: usize,
+        add_rop: bool,
+    ) {
+        assert!(sv_count > sv_changed);
+        // A replacement VCR is created with a single sub_volume that has
+        // a new target and a larger generation, but the other sub_volumes do
+        // not have increased generation numbers, which is not valid
+        // We require that all sub_volumes have bumped generation numbers.
+        let vol_id = Uuid::new_v4();
+        let opts = generic_crucible_opts(vol_id);
+
         let block_size = 512;
         let vol_id = Uuid::new_v4();
         let blocks_per_extent = 10;
         let extent_count = 9;
 
-        let opts = generic_crucible_opts(vol_id);
-
-        let rop = Box::new(VolumeConstructionRequest::Region {
-            block_size,
-            blocks_per_extent,
-            extent_count,
-            opts: opts.clone(),
-            gen: 3,
-        });
+        let rop = if add_rop {
+            let rop = Box::new(VolumeConstructionRequest::Region {
+                block_size,
+                blocks_per_extent,
+                extent_count,
+                opts: opts.clone(),
+                gen: 3,
+            });
+            Some(rop)
+        } else {
+            None
+        };
 
-        let original = VolumeConstructionRequest::Volume {
-            id: vol_id,
-            block_size,
-            sub_volumes: vec![VolumeConstructionRequest::Region {
+        let mut sub_volumes = Vec::new();
```

Comment:
Same suggestion about building this immutably with `(0..sv_count).map(...)`

Reply:
Done in ab9e0223